lib/propolis/src/enlightenment/hyperv/tsc.rs

```diff
@@ -0,0 +1,238 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Support for the Hyper-V reference time enlightenment. See TLFS section 12.7.
+//!
+//! # Theory
+//!
+//! The x86 timestamp counter (TSC) gives system software a high-resolution
+//! performance counter that increments roughly once per processor clock cycle.
+//! The TSC is just a counter and does not return elapsed time in SI units;
+//! instead, readers convert TSC values to elapsed time by dividing the number
+//! of TSC ticks by the TSC frequency to get a number of seconds, which they can
+//! then convert to a reference frequency:
+//!
+//! ```text
+//! elapsed reference units
+//!     = elapsed seconds * (reference units / 1 sec)
+//!     = TSC ticks * (1 / (TSC ticks / 1 sec)) * reference frequency
+//!     = TSC ticks * (1 / TSC frequency) * reference frequency
+//!     = TSC ticks * (reference frequency / TSC frequency)
+//! ```
+//!
+//! (This calculation assumes that the TSC does in fact tick at a constant
+//! frequency. This is often the case on modern processors, but it is not
+//! guaranteed, and system software is expected to check CPUID to see if the CPU
+//! advertises such an "invariant" TSC before doing this kind of calculation.)
+//!
+//! KVM and Linux use nanoseconds as the reference time unit and so have a
+//! reference frequency of 1e9 ticks/sec. Windows and Hyper-V use 100ns units
+//! and a frequency of 1e7 ticks/sec. Because these frequencies are expressed in
+//! cycles per second, using simple integer divisions to convert ticks to
+//! seconds or to scale frequencies will lose all sub-second precision, which
+//! precision is of course the point of having a high-resolution timekeeping
+//! facility. To avoid this problem without having to use floating-point
+//! arithmetic, timekeeping enlightenments usually turn to fixed-point scaling
+//! fractions.
+//!
+//! The general idea is to take an N-bit frequency value and represent it as a
+//! 2N-bit value with an implicit radix point between bits N and N-1, such that
+//! the high N bits of the value are its integer part and the low N bits are its
+//! fractional part. That is, the idea is to re-express the frequency multiplier
+//! term in the conversion above as follows:
+//!
+//! ```text
+//! frequency multiplier
+//!     = reference frequency / TSC frequency
+//!     = ((reference frequency) * 2^N) / TSC frequency) * (1 / 2^N)
+//! ```
+//!
+//! In the present case, N = 64, and the `reference frequency * 2^N` term is a
```

Comment:
Is this just an example? For intel/amd, the size of the fractional number is different (amd is 40 bits with 8int/32 frac, and intel 64 bits with 16 int / 48 frac)

Reply:
Substantially reworked in dce0275 to try to clarify that the 64.64 fixed-point fraction format is chosen by the enlightenment and isn't strictly required to do this kind of math.